Mono.Debugging/Mono.Debugging.Client/SourceLink.cs

```diff
@@ -17,7 +24,25 @@ public SourceLink (string uri, string relativeFilePath)
 
 		public string GetDownloadLocation (string cachePath)
 		{
-			return Path.Combine (cachePath, new Uri (Uri).GetComponents (UriComponents.Path, UriFormat.SafeUnescaped));
+			var uri = new Uri (Uri);
+			return Path.Combine (cachePath, MakeValidFileName (uri));
+		}
+
+		static string MakeValidFileName (Uri uri)
+		{
+			// Remove scheme from uri
+			var text = uri.Host + uri.PathAndQuery;
```

Comment:
We're using a string builder to avoid allocations, but we pay up front for the allocation of this string?